### PR TITLE
Make sbt aware of Dotty

### PR DESCRIPTION
**Pull request by smarter**
This small set of changes, together with the compiler-bridge I wrote
(https://github.com/smarter/dotty-bridge) enables us to compile code
using Dotty in sbt, see https://github.com/smarter/dotty-example-project
for an example.

It'd be great if this could go in 0.13.10 as using a custom sbt is very cumbersome, let me know if there's anything I can do to help with that!

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @smarter,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Typesafe Contributors License Agreement:

[http://www.typesafe.com/contribute/cla](http://www.typesafe.com/contribute/cla)

**Comment by typesafe-tools** [comment]
Can one of the admins verify this patch?

File: compile/src/main/scala/sbt/compiler/RawCompiler.scala [hunk 1]
```diff
@@ -21,14 +21,22 @@ class RawCompiler(val scalaInstance: xsbti.compile.ScalaInstance, cp: ClasspathO
...
+        val process = mainClass.getMethod("process", classOf[Array[String]])
+        process.invoke(null, arguments.toArray)
+        mainClass.getMethod("reporter").invoke(null)
+      }
+    checkForFailure(reporter, arguments.toArray)
   }
   def compilerArguments = new CompilerArguments(scalaInstance, cp)
-  protected def checkForFailure(mainClass: Class[_], args: Array[String]): Unit = {
-    val reporter = mainClass.getMethod("reporter").invoke(null)
+  protected def checkForFailure(reporter: AnyRef, args: Array[String]): Unit = {
```
Comment (dwijnand) [inline compile/src/main/scala/sbt/compiler/RawCompiler.scala:39]: This is a breaking (bincompat) change.

**Comment by dwijnand** [comment]
This is kind of exciting. I'm surprised it's such a small change.

File: compile/src/main/scala/sbt/compiler/RawCompiler.scala (hunk 1 above)
Comment (smarter) [inline compile/src/main/scala/sbt/compiler/RawCompiler.scala:39]: OK, will fix, I have no idea what is and isn't public API :). Is it okay if I push-force my branch or do you prefer code reviews to be addressed in separate commits?

**Comment by smarter** [comment]
This would have been much harder to get in 0.13 without #2197, so thank you @Duhemm  :).

File: compile/src/main/scala/sbt/compiler/RawCompiler.scala (hunk 1 above)
Comment (dwijnand) [inline compile/src/main/scala/sbt/compiler/RawCompiler.scala:39]: Part of 1.0 is making that decision and enforcing it. In these 0.13 days, anything that can be accessed, is public :-/

It's not really been decided, but I'm personally happy with history-rewrites and force pushes.

**Comment by SethTisue** [comment]
perhaps instead of the `isDotty` hack, an actual `SettingKey[Boolean]` called `isDotty`?

**Comment by smarter** [comment]
A setting would be slightly annoying if you want your sbt build to support both Scala and Dotty, you'd have to do `isDotty := scalaVersion.value.startsWith("0.")` yourself, but I'm not opposed to it if people like this better.

**Comment by smarter** [comment]
...Unless the setting default value is `scalaVersion.value.startsWith("0.")` ? I could do this but I have no idea of the proper place to put this setting in the many dimensions of sbt.

**Comment by eed3si9n** [comment]
Settings generally go into `Keys.scala` and `Defaults.scala`. Somewhere around here perhaps? https://github.com/sbt/sbt/blob/v0.13.9/main/src/main/scala/sbt/Defaults.scala#L245

**Comment by smarter** [comment]
Am I allowed to add a `boolean isDotty()` method to the interface `ScalaInstance`?

**Comment by smarter** [comment]
If I make `isDotty` a key then we need a way to access `isDotty` from various classes like `RawCompiler`, the obvious thing to do would be to add an `isDotty` field to the class `ScalaInstance` but this is tricky to do without breaking binary compatibility (I can't change the primary constructor of the class so I'd have to add a mutable field, ...). So I'd rather keep things as-is (no `isDotty` key) for now and revisit this for `1.0` when we'll break binary compatiblity. Anything else that needs to be done before this can be merged?

**Comment by eed3si9n** [comment]
> So I'd rather keep things as-is (no isDotty key) for now and revisit this for 1.0 when we'll break binary compatibility. 

That is a very good point.

> Anything else that needs to be done before this can be merged?

I think we're good here.
